operator/controller/dnsnameresolver-crd/controller.go

```diff
@@ -0,0 +1,132 @@
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	controllerName = "dnsnameresolver_crd_controller"
+)
+
+var (
+	controllerLog = ctrl.Log.WithName(controllerName)
+)
+
+// New creates and returns a controller that creates DNSNameResolver.
+func New(mgr manager.Manager, config Config) (controller.Controller, error) {
+	reconciler := &reconciler{
+		cache:  mgr.GetCache(),
+		client: mgr.GetClient(),
+		config: config,
+	}
+	c, err := controller.New(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, err
+	}
+	clusterNamePredicate := predicate.NewPredicateFuncs(func(o client.Object) bool {
+		crd := manifests.DNSNameResolverCRD().DeepCopy()
+		expectedName := types.NamespacedName{Name: crd.Name}
+		actualName := types.NamespacedName{
+			Name: o.GetName(),
+		}
+		return expectedName == actualName
+	})
+	if err := c.Watch(source.Kind(mgr.GetCache(), &apiextensionsv1.CustomResourceDefinition{}),
+		&handler.EnqueueRequestForObject{}, clusterNamePredicate); err != nil {
+		return nil, err
+	}
+
+	return c, nil
+}
+
+// Config holds all the configuration that must be provided when creating the
+// controller.
+type Config struct {
+	// DependentCaches is a list of caches that are used by Controllers watching DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these caches once
+	// the DNSNameResolver CRD has been created.
+	DependentCaches []cache.Cache
+	// DependentControllers is a list of controllers that watch DNSNameResolver
+	// resources. The dnsnameresolver_crd controller starts these controllers once
+	// the DNSNameResolver CRD has been created and the DependentCaches are started.
+	DependentControllers []controller.Controller
+}
+
+// reconciler handles the actual CRD reconciliation logic in response to
+// events.
+type reconciler struct {
+	config Config
+
+	cache            cache.Cache
+	client           client.Client
+	startCaches      sync.Once
+	startControllers sync.Once
+}
+
+// Reconcile expects request to refer to a CRD and creates or
+// reconciles the DNSNameResolver CRD.
+func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	controllerLog.Info("reconciling", "request", request)
+
+	// Ensure the DNSNameResolver CRD is created.
+	if err := r.ensureDNSNameResolverCRD(ctx); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	// Start the dependent caches and wait for the caches to sync.
+	r.startCaches.Do(func() {
+		var wg sync.WaitGroup
+		for i := range r.config.DependentCaches {
+			cache := &r.config.DependentCaches[i]
+
+			// Start the dependent cache.
+			go func() {
+				if err := (*cache).Start(ctx); err != nil {
```

Comment:
I took inspiration from cluster-ingress-operator code which gets the pointer to the `controller.Controller` interface and then deference it using `(*c)`. Maybe @Miciah can provide some insights as to why this approach was used. It might be because the controller variable in the goroutine it is made sure that the actual controller variable is used rather than a copied value of it.

https://github.com/openshift/cluster-ingress-operator/commit/05c0bcb4409fcd5c4a7b349bc1961ce61dd346f6#diff-74bd64d7a0ece33319f689d27be640577ca56ace22c49b0a07192575c5a8ffb1R104-R114

Reply:
The `cache` variable is created for each `for` iteration so we are not suppose to have a problem with the access to a shared variables. The only thing I can think of is the difference in the receiver object passed to `Start` method. Maybe a value receiver is expected to implement the `Informers` interface.

Reply:
For me it compiles well if `cache` and `controller` variables are of type `interface` and not a pointer to an `interface`:

```diff
$ make build
hack/update-generated-crd.sh
/tmp/coredns-ocp-dnsnameresolver/operator/bin/controller-gen-v0.14.0 rbac:roleName=manager-role crd webhook paths="./..." output:crd:artifacts:config=config/crd/bases
/tmp/coredns-ocp-dnsnameresolver/operator/bin/controller-gen-v0.14.0 object:headerFile="hack/boilerplate.go.txt" paths="./..."
go fmt ./...
go vet ./...
go build -o bin/manager cmd/main.go

$ git diff
diff --git a/operator/controller/dnsnameresolver-crd/controller.go b/operator/controller/dnsnameresolver-crd/controller.go
index 9535e93..d8ceebb 100644
--- a/operator/controller/dnsnameresolver-crd/controller.go
+++ b/operator/controller/dnsnameresolver-crd/controller.go
@@ -92,11 +92,11 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
        r.startCaches.Do(func() {
                var wg sync.WaitGroup
                for i := range r.config.DependentCaches {
-                       cache := &r.config.DependentCaches[i]
+                       cache := r.config.DependentCaches[i]
 
                        // Start the dependent cache.
                        go func() {
-                               if err := (*cache).Start(ctx); err != nil {
+                               if err := cache.Start(ctx); err != nil {
                                        controllerLog.Error(err, "cannot start cache")
                                }
                        }()
@@ -104,7 +104,7 @@ func (r *reconciler) Reconcile(ctx context.Context, request reconcile.Request) (
                        // Wait for the dependent cache to sync.
                        wg.Add(1)
                        go func() {
-                               if started := (*cache).WaitForCacheSync(ctx); !started {
+                               if started := cache.WaitForCacheSync(ctx); !started {
                                        controllerLog.Info("failed to sync cache before starting controllers")
                                }
```